### PR TITLE
[WIP] Improve argument handling for our spec matchers.

### PR DESCRIPTION
**Pull request by troessner**
-> WIP
Fixes the first part of #891 
Speedy review would be awesome since this blocks my rspec refactoring necessary for mutant necessary for #890

### CODE REVIEW

File: lib/reek/spec.rb [hunk 1]
```diff
@@ -83,8 +83,8 @@ module Spec
     #
     # :reek:UtilityFunction
     def reek_of(smell_type,
-                smell_details = {},
-                configuration = Configuration::AppConfiguration.default)
+                smell_details: {},
```
Comment (mvz) [inline lib/reek/spec.rb]: Maybe just `details:`

File: spec/reek/smells/attribute_spec.rb [hunk 2]
```diff
@@ -34,7 +34,7 @@ class Klass
           attr_writer :my_attr
         end
       EOS
-      expect(src).to reek_of(:Attribute, name: 'my_attr')
+      expect(src).to reek_of(:Attribute, smell_details: { name: 'my_attr' })
```
Comment (mvz) [inline spec/reek/smells/attribute_spec.rb]: Hm, I'm not sure I like this extra indirection.

File: lib/reek/spec.rb (hunk 1 above)
Comment (troessner) [inline lib/reek/spec.rb]: Good idea!

File: spec/reek/smells/attribute_spec.rb (hunk 2 above)
Comment (troessner) [inline spec/reek/smells/attribute_spec.rb]: Not even when it's just "details" ? ;)

**Comment by troessner** [comment]
@mvz I removed all changes again and started. Please review and then let's discuss if this is the way to go. If we agree on this (or something else) I'll start to adjust the specs ;)

**Comment by mvz** [comment]
I'll have a look this weekend.

**Comment by mvz** [comment]
I think the interface for `reek_of` and `reek_only_of` is nice as it currently is, and making details a named parameter doesn't improve things for the user of those methods. It only becomes a little awkward if there is also a configuration, but that can be fixed with #606. I can take a stab at that this week I think.

**Comment by troessner** [comment]
>>  It only becomes a little awkward if there is also a configuration, but that can be fixed with #606. I can take a stab at that this week I think.

But right now #606 is blocking multiple bugfixes (because we did a bad job in our branch handling ;) ) and the Reek 4 release.
How about leaving it like it is and fix it in Reek 5?

**Comment by mvz** [comment]
Wait, which fix is blocked by this?

**Comment by mvz** [comment]
.. And which fixes are blocked by #606?

**Comment by troessner** [comment]
>> Wait, which fix is blocked by this?

Let me explain this properly ;)

What we should have done: 

* create a reek-4 release branch
* that gets rebased against master
* so that current bug fixes can still be merged and released

Since we just merged all the breaking changes into master this is not possible.
Right now we quite a few bug fixes in master that we can not release because of that:

* #881 
* #885 
* #890 

And now #893 blocks the release of Reek 4. Which in turn is blocked by #606 
#606 might take a long time or very hard / complicated. We don't know that at the moment. 
#893 on the other hand is conceptually simple once we agreed on an API.
I can finish this off today or tomorrow. And then we could release.

Which is exactly what I'm suggesting ;)

**Comment by mvz** [comment]
Ah, ok, thanks, that clears things up. In that, case, I would say the interface for `reek_of` and friends is good as it is and we shouldn't change it. Which means we can basically go straight to 'Release Reek 4'.

**Comment by troessner** [comment]
Release PR incoming ;)

**Comment by troessner** [comment]
Killing this PR now that #606 is merged.

**Pull request closed without merging**